frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx

```diff
@@ -0,0 +1,126 @@
+  bootOrderEmptyMessage,
+  bootOrderEmptyTitle,
+  deviceKey,
+  deviceLabel,
+  bootOrderAriaLabel,
+} from './constants';
+
+export const BootOrder = ({ devices, setDevices }: BootOrderProps) => {
+  const sources = _.sortBy(devices.filter((device) => device.value.bootOrder), 'value.bootOrder');
+  const options = devices.filter((device) => !device.value.bootOrder);
+  const [isEditMode, setEditMode] = React.useState<boolean>(false);
+
+  // Add a bootOrder to a device by key, item key is "<type>->name>".
+  const addByKey = (key: string): void => {
+    const newDevices = _.cloneDeep(devices);
+
+    const bootOrder = sources.length !== 0 ? sources[sources.length - 1].value.bootOrder + 1 : 1;
+    newDevices.find((device) => deviceKey(device) === key).value.bootOrder = bootOrder;
+
+    setDevices(newDevices);
+    setEditMode(false);
+  };
+
+  // Remove a bootOrder from a device by index.
+  const onDelete = (index: number) => {
+    const newDevices = _.cloneDeep(devices);
+
+    const key = deviceKey(sources[index]);
+    delete newDevices.find((device) => deviceKey(device) === key).value.bootOrder;
+
+    // Reset newDevices bootOrder using the list index.
+    _.filter(newDevices, (device) => device.value.bootOrder).forEach((source, i) => {
+      source.value.bootOrder = i + 1;
+    });
+
+    setDevices(newDevices);
+    setEditMode(false);
+  };
+
+  // Move a source from one index to another.
+  const onMove = (index: number, toIndex: number) => {
+    const unMovedSources = [...sources.slice(0, index), ...sources.slice(index + 1)];
+
+    // Create an ordered copy of the sources.
+    const newSources = _.cloneDeep([
+      ...unMovedSources.slice(0, toIndex),
+      sources[index],
+      ...unMovedSources.slice(toIndex),
+    ]);
+
+    // Reset newSources bootOrder using the list index.
+    newSources.forEach((source, i) => {
+      source.value.bootOrder = i + 1;
+    });
+
+    setDevices([...newSources, ...options]);
+    setEditMode(false);
+  };
+
+  const canAddItem = options.length > 0;
+  const showEmpty = sources.length === 0 && !isEditMode;
+  const dataListID = 'VMBootOrderList';
+
+  return (
+    <>
+      {showEmpty ? (
+        <BootOrderEmpty
+          title={bootOrderEmptyTitle}
+          message={bootOrderEmptyMessage}
+          addItemMessage={addItemMessage}
+          addItemDisabledMessage={addItemDisabledMessage}
+          addItemIsDisabled={options.length === 0}
+          onAdd={() => {
+            setEditMode(true);
+          }}
+        />
+      ) : (
+        <DNDDataList dataListID={dataListID} ariaLabel={bootOrderAriaLabel}>
+          {sources.map((source, index) => (
+            <DNDDataListItem
+              dataListID={dataListID}
+              index={index}
+              onDelete={onDelete}
+              onMove={onMove}
+              key={`${dataListID}-${deviceKey(source)}`}
+            >
+              <Text component={TextVariants.p}>{deviceLabel(source)}</Text>
```

Comment:
shouldn't the `DNDDataListItem` be aria-labelledby this `Text` component rather than the whole list?

Reply:
make sense to me too :-)